src/ngx_http_lua_pipe.c

```diff
@@ -0,0 +1,2473 @@
+                       "lua pipe try to write data %uz pipe:%p", size,
+                       pipe_ctx);
+
+        n = c->send(c, b->pos, size);
+        ngx_log_debug2(NGX_LOG_DEBUG_HTTP, ngx_cycle->log, 0,
+                       "lua pipe write returned %i pipe:%p", n, pipe_ctx);
+
+        if (n >= 0) {
+            b->pos += n;
+
+            if (b->pos == b->last) {
+                b->pos = b->start;
+                b->last = b->start;
+
+                if (!pipe->free_bufs) {
+                    pipe->free_bufs = pipe_ctx->buf_in;
+
+                } else {
+                    pipe->free_bufs->next = pipe_ctx->buf_in;
+                }
+
+                ngx_log_debug1(NGX_LOG_DEBUG_HTTP, ngx_cycle->log, 0,
+                               "lua pipe write done pipe:%p", pipe_ctx);
+                return NGX_OK;
+            }
+
+            continue;
+        }
+
+        /* NGX_ERROR || NGX_AGAIN */
+        break;
+    }
+
+    if (n == NGX_ERROR) {
+        ngx_log_debug1(NGX_LOG_DEBUG_HTTP, ngx_cycle->log, ngx_errno,
+                       "lua pipe write data error pipe:%p", pipe_ctx);
+
+        if (ngx_errno == NGX_EPIPE) {
+            pipe_ctx->err_type = PIPE_ERR_CLOSED;
+
+        } else {
+            pipe_ctx->err_type = PIPE_ERR_SYSCALL;
+            pipe_ctx->pipe_errno = ngx_errno;
+        }
+
+        return NGX_ERROR;
+    }
+
+    return NGX_AGAIN;
+}
+
+
+ssize_t
+ngx_http_lua_ffi_pipe_proc_write(ngx_http_request_t *r,
+    ngx_http_lua_ffi_pipe_proc_t *proc, const u_char *data, size_t len,
+    u_char *errbuf, size_t *errbuf_size)
+{
+    int                                 rc;
+    ngx_buf_t                          *b;
+    ngx_msec_t                          timeout;
+    ngx_chain_t                        *cl;
+    ngx_event_t                        *wev;
+    ngx_http_lua_ctx_t                 *ctx;
+    ngx_http_lua_pipe_t                *pipe;
+    ngx_http_lua_co_ctx_t              *wait_co_ctx;
+    ngx_http_lua_pipe_ctx_t            *pipe_ctx;
+
+    rc = ngx_http_lua_pipe_get_lua_ctx(r, &ctx, errbuf, errbuf_size);
+    if (rc != NGX_OK) {
+        return rc;
+    }
+
+    ngx_log_debug2(NGX_LOG_DEBUG_HTTP, r->connection->log, 0,
+                   "lua pipe write process:%p pid:%P", proc, proc->_pid);
+
+    pipe = proc->pipe;
+    if (pipe == NULL || pipe->closed) {
+        *errbuf_size = ngx_snprintf(errbuf, *errbuf_size, "closed") - errbuf;
+        return NGX_ERROR;
+    }
+
+    if (pipe->stdin_ctx == NULL) {
+        if (ngx_http_lua_pipe_init_ctx(&pipe->stdin_ctx, pipe->stdin_fd,
+                                       pipe->pool, errbuf,
+                                       errbuf_size)
+            != NGX_OK)
+        {
+            return NGX_ERROR;
+        }
+
+    } else {
+        pipe->stdin_ctx->err_type = 0;
+    }
+
+    pipe_ctx = pipe->stdin_ctx;
+    if (pipe_ctx->c == NULL) {
+        *errbuf_size = ngx_snprintf(errbuf, *errbuf_size, "closed") - errbuf;
```

Comment:
We still override the errors set by `ngx_http_lua_pipe_init_ctx` here. Let's either not set the error buffer there, or follow the suggestion elaborated in my previous comment here. The latter provides more precise error messages (`"closed"` can be misleading if, e.g., we are in fact hitting the `"no memory"` branch of `ngx_http_lua_pipe_init_ctx`.

Reply:
@spacewander Ping here. This is my last concern for this PR.

Reply:
> We still override the errors set by `ngx_http_lua_pipe_init_ctx` here.

Err... I am quite confused. When `ngx_http_lua_pipe_init_ctx` sets an error, it will return `NGX_ERROR`. Then the `ngx_http_lua_ffi_pipe_proc_write` will return from there:
https://github.com/openresty/lua-nginx-module/blob/36bb55ff7cef202de6ecbbc8cfb39c9b079defd0/src/ngx_http_lua_pipe.c#L1802

So this line:
https://github.com/openresty/lua-nginx-module/blob/36bb55ff7cef202de6ecbbc8cfb39c9b079defd0/src/ngx_http_lua_pipe.c#L1811
won't be hit, and errors won't be overridden.

Reply:
Err, I guess I am quite confused myself... marking this as resolved!